File: problems/007/007.cpp
#include<cstdint>
#include<cstdlib>
#include<ctime>
#include<iostream>
#include<unordered_map>

using namespace std;

__attribute__((always_inline))
inline uint64_t f() {
  int i = 0, n = 10000;
  uint64_t q = 1;
  unordered_map<uint64_t, uint64_t> map;

  while (true) {
    q += 1;

    uint64_t p = map[q];

    if (p == 0) {
      map[q * q] = q;

      if (i == n) {
        return q;
      } else {
        i += 1;
      }
    } else {
      uint64_t x = p + q;

      map[q] = 0;

      while (map[x] != 0) {
        x += p;
      }

      map[x] = p;
    }
  }
}

uint64_t toNs(timespec &ts) {
  return (uint64_t)ts.tv_sec * 1000000000lu + (uint64_t)ts.tv_nsec;
}

int main(int argc, char *argv[]) {
  if (argc == 1) {
    cout << f() << endl;
  } else if (argc == 2) {
    timespec start, end;
    uint64_t iters = atol(argv[1]);

    clock_gettime(CLOCK_MONOTONIC, &start);
    for (uint64_t i = 0; i < iters; i++) {
      asm("");
      f();
    }
    clock_gettime(CLOCK_MONOTONIC, &end);

    cout << toNs(end) - toNs(start) << endl;
  }
}

File: problems/007/007.rb
def f
  i, map, n, q = 0, {}, 10_000, 1

  loop do
    q += 1

    p = map.delete(q)

    if p.nil?
      map[q * q] = q

      if i == n
        return q
      else
        i += 1
      end
    else
      x = p + q

      x += p while map.key?(x)

      map[x] = p
    end
  end
end

case ARGV.length
when 0
  puts f
when 1
  iters = ARGV.first.to_i

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  (1..iters).each do ||
    f
  end
  end_ = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  elapsed = ((end_ - start) * 1e9).to_i

  puts elapsed
end
